### PR TITLE
small fix to make fake viz only detect objects in front of the camera

### PR DESCRIPTION
**Pull request by SammyRamone**
fix for https://github.com/bit-bots/bitbots_misc/issues/17

### CODE REVIEW

**Comment by SammyRamone** [comment]
tested it and it worked

**Comment by SammyRamone** [comment]
> I think it would be better to check p_pixel[z] to be positive, because it might happen for some head modes to detect objects behind the robot. it would also be more logically correct and more efficient, as it can be done before the matrix multiplication

The p_pixel[z] is not possible to use since it is always 1
But you are correct, the ball pose is in base footprint frame (I thought that its in camera frame). I will transform it to camera frame and then it should be okay

**Comment by MonsterUntermBett** [comment]
> > I think it would be better to check p_pixel[z] to be positive, because it might happen for some head modes to detect objects behind the robot. it would also be more logically correct and more efficient, as it can be done before the matrix multiplication
> 
> The p_pixel[z] is not possible to use since it is always 1
> But you are correct, the ball pose is in base footprint frame (I thought that its in camera frame). I will transform it to camera frame and then it should be okay

Or you could just do the check right after getting the transforms, which already are in camera_optical_frame. This would save useless calculations.

**Comment by SammyRamone** [comment]
I did some changes. I can not use the pose in camera_optical_frame, since the z axis is not correct there, I have to use camera frame. Still, I can do the first check before performing further transformations.
I didn't test the changes, since I can't start the simulation on the computer that I'm currently on, but I think that this could solve the problem.

File: bitbots_bringup/scripts/fake_vis_in_sim.py [hunk 1]
```diff
@@ -59,66 +59,75 @@ def state_update(state_msg):
     ball_pose_stamped.header.stamp = rospy.Time.now()
     ball_pose_stamped.header.frame_id = "map"
     ball_pose_stamped.pose = ball_pose
+    ball_in_camera_frame = tf_buffer.transform(ball_pose_stamped, "camera", timeout=rospy.Duration(0.5))
```
Comment (jgueldenstein) [inline bitbots_bringup/scripts/fake_vis_in_sim.py]: why is this in the camera frame and not in the camera optical frame

File: bitbots_bringup/scripts/fake_vis_in_sim.py [hunk 2]
```diff
@@ -59,66 +59,75 @@ def state_update(state_msg):
...
-                                                timeout=rospy.Duration(0.5)).pose.position
-            lp = True
 
-        p = [right_post.pose.position.x, right_post.pose.position.y, right_post.pose.position.z]
-        k = np.reshape(cam_info.K, (3, 3))
-        p_pixel = np.matmul(k, p)
-        p_pixel = p_pixel * (1 / p_pixel[2])
+        left_post = deepcopy(goal_pose_stamped)
+        left_post.pose.position.y += 1.35
+        left_post_in_camera_frame = tf_buffer.transform(left_post, "camera", timeout=rospy.Duration(0.5))
```
Comment (jgueldenstein) [inline bitbots_bringup/scripts/fake_vis_in_sim.py]: this should also be camera optical or better cam_info.header.frame_id

File: bitbots_bringup/scripts/fake_vis_in_sim.py [hunk 3]
```diff
@@ -59,66 +59,75 @@ def state_update(state_msg):
...
+        if left_post_in_camera_frame.pose.position.z >= 0:
+            left_post = tf_buffer.transform(left_post, cam_info.header.frame_id,
+                                                    timeout=rospy.Duration(0.5))
+            p = [left_post.pose.position.x, left_post.pose.position.y, left_post.pose.position.z]
+            k = np.reshape(cam_info.K, (3, 3))
+            p_pixel = np.matmul(k, p)
+            p_pixel = p_pixel * (1 / p_pixel[2])
+            lp = False
+            if p_pixel[0] > 0 and p_pixel[0] <= cam_info.width and p_pixel[1] > 0 and p_pixel[1] <= cam_info.height\
+                    and left_post.pose.position.z > 0:
```
Comment (jgueldenstein) [inline bitbots_bringup/scripts/fake_vis_in_sim.py]: the z component of the position is checked already in line 100

File: bitbots_bringup/scripts/fake_vis_in_sim.py [hunk 4]
```diff
@@ -59,66 +59,75 @@ def state_update(state_msg):
...
+            right_post = tf_buffer.transform(right_post, cam_info.header.frame_id,
+                                                    timeout=rospy.Duration(0.5))
+            p = [right_post.pose.position.x, right_post.pose.position.y, right_post.pose.position.z]
+            k = np.reshape(cam_info.K, (3, 3))
+            p_pixel = np.matmul(k, p)
+            p_pixel = p_pixel * (1 / p_pixel[2])
+
+            rp = False
+            if p_pixel[0] > 0 and p_pixel[0] <= cam_info.width and p_pixel[1] > 0 and p_pixel[1] <= cam_info.height\
+                and right_post.pose.position.z:
```
Comment (jgueldenstein) [inline bitbots_bringup/scripts/fake_vis_in_sim.py]: this line does not really make sense

**Comment by jgueldenstein** [comment]
> I did some changes. I can not use the pose in camera_optical_frame, since the z axis is not correct there, I have to use camera frame. Still, I can do the first check before performing further transformations.

what you are talking about is the position after the multiplication with the projection matrix, before that, the camera_optical_frame is just a regular frame and any position can be represented in it,

**Comment by SammyRamone** [comment]
i deleted the script and moved it to another package (see https://github.com/bit-bots/humanoid_league_visualization/pull/6)
this pull request should still be accepted to remove the script

**Comment by jgueldenstein** [comment]
simulator_teamplayer.launch needs to be adapted

**Comment by SammyRamone** [comment]
> simulator_teamplayer.launch needs to be adapted

correct, I fixed that

**Pull request merged**